mrpeek: (time series) plot tool

maxpietsch opened the pull request:
At the cross-hair position, key `p` produces a 2D plot of intensity values at that location across a specified dimension.

Todo:
- [x] dimension selection, here it is the current slice direction
- [x] show focus, changing the focus updates the plot
- [x] nan handling
- [x] nicer plot: add some padding

![image](https://user-images.githubusercontent.com/10046944/85207668-b4b54b80-b32a-11ea-859f-c756e68255bd.png)

![image](https://user-images.githubusercontent.com/10046944/85207674-c0087700-b32a-11ea-8c4e-37cf2d0b9d9f.png)


![image](https://user-images.githubusercontent.com/10046944/85209228-7756bb00-b336-11ea-8d04-bfee0ffbd544.png)


![image](https://user-images.githubusercontent.com/10046944/85209233-7d4c9c00-b336-11ea-8e85-89153f8c6064.png)

maxpietsch committed:
mrpeek: added plot tool

jdtournier commented:
😍

If only we had that in `mrview` itself... 😭

maxpietsch commented:
Just port ODF and fixel rendering and we can deprecate `mrview`.

jdtournier commented:
😭

jdtournier commented:
Not all that far fetched...
https://github.com/saitoha/SDL1.2-SIXEL/commit/ab3fccac6e34260a617be511bd8c2b2beae41952

maxpietsch committed:
mrpeek: add plot focus and lines between data

maxpietsch committed:
mrpeek: plot current position, deactivate colour bar control

dchristiaens commented:
> Just port ODF and fixel rendering and we can deprecate `mrview`.

I would have thought these would be priority over time series plotting. Not planning on switching to fMRI are you? 😄

jdtournier commented:
Looks great! I reckon this would work best with the `query_functionality` branch: `p` to prompt the question "plot along axis:", user enters axis, and that's what's displayed. Empty answer deactivates. 

I've not had a chance to check this out yet, but is the plot shown to the side of the image? And updated on change of focus?

maxpietsch commented:
No no interest in fMRI time series, but in venetian blind artefacts, signal decay curves, relaxation curves, 🦓,... 

P replaces the image with the plot pane but we could use the ortho view functionality to show the slice and the plot. The plot updates with the focus, the crosshair looks different for the current focus position and is updated when modified (see first post).

maxpietsch committed:
Merge branch 'master' into mrpeek_plot

maxpietsch committed:
mrpeek: correct indexing with nonfinite values

maxpietsch committed:
merge query_functionality into mrpeek_plot

maxpietsch committed:
mrpeek: interactive plot axis selection

maxpietsch committed:
mrpeek: merge focus_control and fix plot

maxpietsch commented:
After a quite painful merge with #31, I think it's ready for prime time:

![image](https://user-images.githubusercontent.com/10046944/85296479-c10fe480-b4a1-11ea-902f-a38cb199e602.png)

![image](https://user-images.githubusercontent.com/10046944/85298276-2f55a680-b4a4-11ea-87ff-5130d607804c.png)

![image](https://user-images.githubusercontent.com/10046944/85298407-5318ec80-b4a4-11ea-8ee7-07b9e56dcdf4.png)

maxpietsch committed:
merge master into mrpeek_plot

jdtournier committed:
Merge remote-tracking branch 'github/master' into mrpeek_plot

jdtournier committed:
show plot below main image

Along with assorted changes to restore image interactions previously
ignored while plot was shown.

jdtournier commented:
OK, I've just pushed some changes to have the plot show _below_ the main image, which looks like this:

![Screenshot from 2020-06-23 19-42-29](https://user-images.githubusercontent.com/3221000/85444170-fb54b100-b589-11ea-9f72-d5eda3937b80.png)

This means the image retains full interactivity, so I've gone and restored all of the actions previously disabled while in plot mode. 

What do you guys reckon? 

Points to note: 

- your render might differ from mine, which I think stems from the handling of the newline character. Technically, in raw mode it shouldn't have any effect. But there's a good chance it might on some terminals. But this seems to render the same on minTTY and mlTerm on my end.

- related to the above, avoid using newline or `std::endl`, it won't do what you expect. Use `VT::move_cursor(VT::Down,1)` instead, followed by `VT::CarriageReturn`.

jdtournier commented:
One more point: in due course, we should go through the code and change all writes to `std::cout` with appends to a string buffer, and condense the screen update to a single write to `std::cout` & flush. This might help to reduce some of the flicker, which is noticeable on a slower connection.

jdtournier committed:
restore y axis labels, and tweak plot status line

jdtournier commented:
@maxpietsch: sorry, just noticed I'd nuked your _y_ axis range labels... Just fixed that, this is what it now looks like:

![Screenshot from 2020-06-23 20-02-31](https://user-images.githubusercontent.com/3221000/85446624-92226d00-b58c-11ea-8c7f-01405fd54c14.png)

maxpietsch commented:
Hmm looks good in the screenshots but in iTerm2, the bottom line of the plot is cut off and the plot crashes:

![image](https://user-images.githubusercontent.com/10046944/85447961-76bc5f80-b596-11ea-98aa-618c819b44d4.png)

maxpietsch committed:
mrpeek: fix plot tool

jdtournier commented:
OK, did that fix it? Does it display as it should now?

maxpietsch committed:
mrpeek: add newline to avoid text and sixel overlap in iterm2

maxpietsch commented:
The crash might have been a botched merge on my side which I fixed in 343baeb. The line ist still cropped. I had to add another`VT::move_cursor(VT::Down,1)` before the carriage return (855b372) to get this:

![image](https://user-images.githubusercontent.com/10046944/85450936-9f922400-b599-11ea-94bc-72b9ef14961f.png)


EDIT here is mlterm:
![image](https://user-images.githubusercontent.com/10046944/85452407-2398db80-b59b-11ea-95bf-4e828d8743c0.png)

jdtournier commented:
That's a bummer... on my system, that extra line shift is not required, and doesn't look great. Seems iTerm2 handles the position of the cursor after a sixel image differently from mlTerm and minTTY...

maxpietsch commented:
Shall we move the index text above the sixel? Not great but better than overlap or extra line ¯\_(ツ)_/¯ There would be extra space between image and plot and between plot and lower y axis label but that's better than for the main image.

jdtournier committed:
add function to get cursor position and test newline after sixel

Also adds a function to get the terminal size, which may be useful in
future.

jdtournier commented:
Just added a quick test to see whether we can detect whether we need a newline after the sixel plot or not. The relevant test is [here](https://github.com/MRtrix3/mrpeek/pull/27/commits/8346ff4beeb4c77dc6a1543b0c5568ff82c045e8#diff-268469edc6bf7d57a39f7be7257b19aaR235). If that doesn't work, you could also try testing whether `x==1` on that line (on my system, `x` is 2 at that point, the cursor will have moved down one row after the sixel dump).

jdtournier committed:
clean up need for newline after sixel handling

This is now all internally handled by the sixel encoder, with the
expectation that the output should match across terminals regardless of
whether or not a newline is inserted after the sixel.

jdtournier commented:
OK, I've just modified the test to actually check the row number - seems more direct somehow given the issue. 
I've also moved all the handling of the issue within the sixel encoder, so it shouldn't need special handling outside of that - the results should match regardless of terminal. :crossed_fingers:

maxpietsch commented on src/sixel.h:
```diff
@@ -173,6 +177,8 @@ namespace MR {
           // replace last character from $ (carriage return) to '-' (newline):
           out.back() = '-';
+          if (need_newline_after_sixel)
+            out += VT::move_cursor (VT::Down,1) + VT::CarriageReturn;
```
this line breaks sixels in iTerm2, I get text to the terminal instead of the image

jdtournier replied:
Sorry, my mistake... But it looks like you've already figured it out?

maxpietsch committed:
mrpeek: fix sixel encode

maxpietsch commented:
Works:

iTerm2
![image](https://user-images.githubusercontent.com/10046944/85470572-85634080-b5af-11ea-8ebb-9699a9bbadce.png)

mlTerm
![image](https://user-images.githubusercontent.com/10046944/85470868-eab73180-b5af-11ea-8269-73f34bd7c454.png)

jdtournier committed:
remove dangling VAR()

jdtournier commented:
:tada:

jdtournier commented:
Good to go? Feel free to merge!

maxpietsch commented:
🍾
